### ISSUE TITLE
Changelogs for packages aren't documented

### ISSUE THREAD

**Issue by geekosaur** [haskell/cabal#10474]
**What is wrong with the docs?**
According to a user in `#hackage:matrix.org`, nowhere is it documented that Hackage requires packages to have a changelog file, nor what this involves.

**Additional context**
```
[24 09:40:50]  <mauke> the other day hackage complained that my package doesn't have a changelog
[24 09:41:03]  <mauke> what is a changelog, where does hackage look for it, and where is this documented?
[24 09:52:13]  <haskellbridge> <H​écate> mauke: https://en.wikipedia.org/wiki/Changelog
[24 09:52:43]  <haskellbridge> <H​écate> CHANGELOG.md is a good place to have your changelog
[24 09:53:36]  <haskellbridge> <H​écate> and I believe it has to be put in https://cabal.readthedocs.io/en/stable/cabal-package-description-file.html#pkg-field-extra-doc-files
[24 09:53:47]  <haskellbridge> <H​écate> mauke: run "cabal check" I think it'll tell you more
[24 09:54:17]  <mauke> cabal check just says "No errors or warnings could be found in the package."
[24 10:34:37]  <haskellbridge> <m​auke> I can't find changelog anywhere in the cabal documentation. Or hackage. 
[24 10:54:58]  <haskellbridge> <f​-a> mauke: please tell us your "cabal --version"
[24 10:59:45]  <haskellbridge> <f​-a> This is the pr #8497 (https://github.com/haskell/cabal/pull/8747) (alas, missing a changelog itself!)
```

**Comment by philderbeast** [comment]
I used `cabal init --non-interactive` and then deleted the `CHANGELOG.md` and its reference in the package description `.cabal` file. Rerunning `cabal sdist` works the same as it did before the deletion, writing the tarball. The `sdist` command is not checking that there is a change log.

**Comment by ffaf1** [comment]
I tried to upload a package candidate without changelog to Hackage, it works:

```
f@x270:~/download/lentil-1.5.8.0$ cabal upload /home/f/download/lentil-1.5.8.0/dist-newstyle/sdist/lentil-1.5.8.0.tar.gz
Warning: this is a debug build of cabal-install with assertions enabled.
Uploading
/home/f/download/lentil-1.5.8.0/dist-newstyle/sdist/lentil-1.5.8.0.tar.gz...
Package successfully uploaded as candidate. You can now preview the result at
'https://hackage.haskell.org/package/lentil-1.5.8.0/candidate'. To publish the
candidate, use 'cabal upload --publish'.
f@x270:~/download/lentil-1.5.8.0$ ls
dist-newstyle  doc  lentil.cabal  LICENSE  README  Setup.hs  src  stack.yaml  test
```

I ask the user for more informations.

**Comment by ffaf1** [comment]
More from the user.
If you upload a package without changelog, Hackage shows the field with “None Available” in bold/red.

```
<user> ah, yes. that was the red notice I remember
<user> it caught my eye because I was following the cabal and hackage documentation to make sure I'd added all the required <user> fields, and then suddenly this "Change log" property shows up in red
<user> when it has never been mentioned before
```

I wonder how suggestive `cabal check` should be.
Pestering the user to add a changelog file will lead to a number of changelog being there, but not actually being updated.

**Comment by ulysses4ever** [comment]
I don't believe changelogs are mandatory on Hackage (e.g. here's a package without one uploaded in 2020: https://hackage.haskell.org/package/BNFC-meta).

**Comment by ulysses4ever** [comment]
> I wonder how suggestive cabal check should be.
> Pestering the user to add a changelog file will lead to a number of changelog being there, but not actually being updated.

Well, pestering is not the right approach, of course. But helpfully suggest to consider adding one so that the users have an idea how this release progresses from the previous one would be good imo.

**Comment by geekosaur** [comment]
While this is all cogent, it's missing the user's (and my, cf. the title I chose for this issue) point: **it's not documented**.

**Comment by ffaf1** [comment]
> While this is all cogent, it's missing the user's (and my, cf. the title I chose for this issue) point: it's not documented.

I hate the “not our problem” attitude, especially in tightly integrated tools, but is there anything we should do here, *doc-wise*?
This is something Hackage does, something that is presentational and will not prevent you from uploading a package.

`cabal init` also already creates a `CHANGELOG.md` plus `extra-doc-file` entry in the manifest, so from an UX POV you need to actively remove those, not something you stumble upon by chance.

That would be different of course if we were to create a new warning, then documentation would be necessary.

**Comment by ffaf1** [comment]
> Some documentation about changelogs, including something about the way they're inferred

Alas they are inferred in two different ways, cabal does one thing and Hackage does a similar (but not equivalent) thing.

I note there is a (tiny, not fully explained) changelog reference in the [`cabal check` docs](https://cabal.readthedocs.io/en/stable/cabal-commands.html#cabal-check):

> `no-docs`: missing expected documentation files (changelog).

(this is only tangent to this ticket tho, i.e. the case where changelos is present in the working directory but not in the `.cabal` file).

I will make a patch to expand this bit, for sure.

**Comment by mauke** [comment]
> `cabal init` also already creates a `CHANGELOG.md` plus `extra-doc-file` entry in the manifest, so from an UX POV you need to actively remove those, not something you stumble upon by chance.

That's a relatively recent feature. As far as I can tell, creation of a changelog was only added to cabal-install in 4d6175ccb3b17e0 (from 2015, probably v1.24?). My package is older than that.

(For extra irony, the cabal-install changelog does not mention when `cabal init` started generating changelogs.)
